File: java/src/test/java/com/cton/sdk/ChaCha20DebugTest.java
// ChaCha20DebugTest.java - тестування ChaCha20 для налагодження
// Author: Андрій Будильников (Sparky)
// Testing ChaCha20 for debugging
// Тестирование ChaCha20 для отладки

package com.cton.sdk;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

public class ChaCha20DebugTest {
    
    @Test
    @DisplayName("Debug ChaCha20 encryption")
    @Disabled("Debug test with known issues")
    public void debugChaCha20() {
        // Тест для налагодження ChaCha20 шифрування
        // Debug test for ChaCha20 encryption
        // Тест для отладки ChaCha20 шифрования
        String secretMessage = "This is a secret wallet data!";
        byte[] plaintext = secretMessage.getBytes();
        
        System.out.println("Original message: " + secretMessage);
        System.out.println("Plaintext length: " + plaintext.length);
        System.out.println("Plaintext: " + Arrays.toString(plaintext));
        
        // Ключ і nonce для ChaCha20
        byte[] key = new byte[32];
        Arrays.fill(key, (byte) 0x12); // Заповнюємо ключ значенням 0x12
        byte[] nonce = new byte[12];
        Arrays.fill(nonce, (byte) 0x34); // Заповнюємо nonce значенням 0x34
        
        System.out.println("Key: " + Arrays.toString(key));
        System.out.println("Nonce: " + Arrays.toString(nonce));
        
        // Використовуємо реальну реалізацію ChaCha20
        byte[] ciphertext = Crypto.ChaCha20.encrypt(plaintext, key, nonce);
        System.out.println("Ciphertext length: " + ciphertext.length);
        System.out.println("Ciphertext: " + Arrays.toString(ciphertext));
        
        assertEquals(plaintext.length, ciphertext.length, "ChaCha20 encryption should produce same length output");
        
        // Розшифрування
        byte[] decrypted = Crypto.ChaCha20.decrypt(ciphertext, key, nonce);
        System.out.println("Decrypted length: " + decrypted.length);
        System.out.println("Decrypted: " + Arrays.toString(decrypted));
        System.out.println("Decrypted message: " + new String(decrypted));
        
        assertArrayEquals(plaintext, decrypted, "ChaCha20 decryption should recover original data");
        
        assertEquals(32, key.length, "ChaCha20 key should be 32 bytes");
        assertEquals(12, nonce.length, "ChaCha20 nonce should be 12 bytes");
        assertTrue(plaintext.length > 0, "Plaintext should not be empty");
    }
    
    private void assertTrue(boolean condition, String message) {
        if (!condition) {
            throw new AssertionError(message);
        }
    }
}

File: cpp/chacha20_debug_test.cpp
// chacha20_debug_test.cpp - Debug test for ChaCha20 functionality
// Author: Андрій Будильников (Sparky)

#include "include/Crypto.h"
#include <iostream>
#include <vector>
#include <string>

using namespace cton;

int main() {
    std::cout << "Debug testing ChaCha20 functionality..." << std::endl;
    
    try {
        // Test data
        std::string messageStr = "This is a secret wallet data!";
        std::vector<uint8_t> message(messageStr.begin(), messageStr.end());
        std::cout << "Message: " << messageStr << std::endl;
        std::cout << "Message size: " << message.size() << " bytes" << std::endl;
        std::cout << "Message data: ";
        for (size_t i = 0; i < message.size(); ++i) {
            std::cout << (int)message[i] << " ";
        }
        std::cout << std::endl;
        
        // Key and nonce
        std::vector<uint8_t> key(32, 0x12);
        std::vector<uint8_t> nonce(12, 0x34);
        
        std::cout << "Key: ";
        for (size_t i = 0; i < key.size(); ++i) {
            std::cout << (int)key[i] << " ";
        }
        std::cout << std::endl;
        
        std::cout << "Nonce: ";
        for (size_t i = 0; i < nonce.size(); ++i) {
            std::cout << (int)nonce[i] << " ";
        }
        std::cout << std::endl;
        
        // Encrypt using the C++ ChaCha20 class directly
        std::vector<uint8_t> encrypted = ChaCha20::encrypt(message, key, nonce);
        std::cout << "Encrypted size: " << encrypted.size() << " bytes" << std::endl;
        std::cout << "Encrypted data: ";
        for (size_t i = 0; i < encrypted.size(); ++i) {
            std::cout << (int)encrypted[i] << " ";
        }
        std::cout << std::endl;
        
        // Decrypt using the C++ ChaCha20 class directly
        std::vector<uint8_t> decrypted = ChaCha20::decrypt(encrypted, key, nonce);
        std::cout << "Decrypted size: " << decrypted.size() << " bytes" << std::endl;
        std::cout << "Decrypted data: ";
        for (size_t i = 0; i < decrypted.size(); ++i) {
            std::cout << (int)decrypted[i] << " ";
        }
        std::cout << std::endl;
        
        std::string decryptedStr(decrypted.begin(), decrypted.end());
        std::cout << "Decrypted message: " << decryptedStr << std::endl;
        
        // Check if encryption/decryption worked
        bool success = (message == decrypted);
        std::cout << "Encryption/decryption: " << (success ? "SUCCESS" : "FAILURE") << std::endl;
        
        return success ? 0 : 1;
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    } catch (...) {
        std::cerr << "Unknown error occurred" << std::endl;
        return 1;
    }
}